# Тема “Обучение с учителем”


## Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


#### Импортируйте библиотеки pandas и numpy.

In [42]:
import pandas as pd
import numpy as np

#### Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn.

In [43]:
from sklearn.datasets import load_boston

#### Создайте датафреймы X и y из этих данных.

In [44]:
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [55]:
X = pd.DataFrame(boston['data'], columns=boston['feature_names'])
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [66]:
y = pd.DataFrame(boston['target'], columns=['price'])
y.head()
#y.info()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

#### Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [47]:
from sklearn.model_selection import train_test_split

In [70]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train.head()

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
5     0.02985  0.0   2.18   0.0  0.458  6.430  58.7  6.0622   3.0  222.0   
116   0.13158  0.0  10.01   0.0  0.547  6.176  72.5  2.7301   6.0  432.0   
45    0.17142  0.0   6.91   0.0  0.448  5.682  33.8  5.1004   3.0  233.0   
16    1.05393  0.0   8.14   0.0  0.538  5.935  29.3  4.4986   4.0  307.0   
468  15.57570  0.0  18.10   0.0  0.580  5.926  71.0  2.9084  24.0  666.0   

     PTRATIO       B  LSTAT  
5       18.7  394.12   5.21  
116     17.8  393.30  12.04  
45      17.9  396.90  10.21  
16      21.0  386.85   6.58  
468     20.2  368.74  18.13

In [71]:
X_test.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
173  0.09178   0.0   4.05   0.0  0.510  6.416  84.1  2.6463   5.0  296.0   
274  0.05644  40.0   6.41   1.0  0.447  6.758  32.9  4.0776   4.0  254.0   
491  0.10574   0.0  27.74   0.0  0.609  5.983  98.8  1.8681   4.0  711.0   
72   0.09164   0.0  10.81   0.0  0.413  6.065   7.8  5.2873   4.0  305.0   
452  5.09017   0.0  18.10   0.0  0.713  6.297  91.8  2.3682  24.0  666.0   

     PTRATIO       B  LSTAT  
173     16.6  395.50   9.04  
274     17.6  396.90   3.53  
491     20.1  390.11  18.07  
72      19.2  390.91   5.52  
452     20.2  385.09  17.27

In [72]:
y_train.head()

price
5     28.7
116   21.2
45    19.3
16    23.1
468   19.1

In [73]:
y_test.head()

price
173   23.6
274   32.4
491   13.6
72    22.8
452   16.1

#### Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [58]:
from sklearn.linear_model import LinearRegression

In [59]:
lr = LinearRegression()

#### Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [63]:
lr.fit(X_train, y_train)

LinearRegression()

In [75]:
y_pred = lr.predict(X_test)

y_pred.shape

(152, 1)

#### Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [77]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [79]:
from sklearn.metrics import r2_score

In [81]:
r2_score(check_test["y_pred"], check_test["y_test"])

0.6693702691495608

## Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.


#### Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42. 

In [83]:
from sklearn.ensemble import RandomForestRegressor

In [84]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

#### Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма. 

In [96]:
model.fit(X_train, y_train.values[:,0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [101]:
y_pred1 = model.predict(X_test)

y_pred1.shape

(152,)

#### Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания. Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [102]:
check_test1 = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred1": y_pred.flatten(),
})

check_test1.head(10)

y_test    y_pred1
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
76     20.0  21.832284
316    17.8  19.895747
140    14.0  14.754118
471    19.6  21.240835
500    16.8  20.898658

In [104]:
r2_score(check_test1["y_pred1"], check_test["y_test"])

0.8479049999699443

#### Сравнение результатов
В первом случае R2=0.6693702691495608, во втором R2=0.8479049999699443, вторая модель работает лучше.